docs/architecture/adr-030-consensus-refactor.md

```diff
@@ -0,0 +1,152 @@
+# ADR 030: Consensus Refactor
+
+## Context
+
+One of the biggest challenges this project faces is to proof that the
+implementations of the specifications are correct, much like we strive to
+formaly verify our alogrithms and protocols we should work towards high
+confidence about the correctness of our program code. One of those is the core
+of Tendermint - Consensus - which currently resides in the `consensus` package.
+Over time there has been high friction making changes to the package due to the
+algorithm being scattered in a side-effectful container (the current
+`ConsensusState`). In order to test the algorithm a large object-graph needs to
+be set up and even than the non-deterministic parts of the container makes will
+prevent high certainty. Where ideally we have a 1-to-1 representation of the
+[spec](https://github.com/tendermint/spec), ready and easy to test for domain
+experts.
+
+Addresses:
+
+- [#1495](https://github.com/tendermint/tendermint/issues/1495)
+- [#1692](https://github.com/tendermint/tendermint/issues/1692)
+
+## Decision
+
+To remedy these issues we plan a gradual, non-invasive refactoring of the
+`consensus` package. Starting of by isolating the consensus alogrithm into
+a pure function and a finite state machine to address the most pressuring issue
+of lack of confidence. Doing so while leaving the rest of the package in tact
+and have follow-up optional changes to improve the sepration of concerns.
+
+### Implementation changes
+
+The core of Consensus can be modelled as a function with clear defined inputs:
+
+* `State` - data container for current round, height, etc.
+* `Event`- significant events in the network
+
+producing clear outputs;
+
+* `State` - updated input
+* `Message` - signal what actions to perform
+
+```go
+type Event int
+
+const (
+	EventUnknown Event = iota
+	EventProposal
+	Majority23PrevotesBlock
+	Majority23PrecommitBlock
+	Majority23PrevotesAny
+	Majority23PrecommitAny
+	TimeoutNewRound
+	TimeoutPropose
+	TimeoutPrevotes
+	TimeoutPrecommit
+)
+
+type Message int
+
+const (
+	MeesageUnknown Message = iota
+	MessageProposal
+	MessageVotes
+	MessageDecision
+)
+
+type State struct {
+	height      uint64
+	round       uint64
+	step        uint64
+	lockedValue interface{} // TODO: Define proper type.
+	lockedRound interface{} // TODO: Define proper type.
+	validValue  interface{} // TODO: Define proper type.
+	validRound  interface{} // TODO: Define proper type.
+	// From the original notes: valid(v)
+	valid       interface{} // TODO: Define proper type.
+	// From the original notes: proposer(h, r)
+	proposer    interface{} // TODO: Define proper type.
+}
+
+func Consensus(Event, State) (State, Message) {
```

Comment:
I think it's a bit weird to call this "Consensus".  It's a very small part of consensus... and "step" is overloaded... maybe "ConsensusIota". 

While agree that ConsensusState can be split into ConsensusExecutor and ConsensusState, I'd still want to keep the exact same method signature:

```go
func (cs *ConsensusState) enterNewRound(height int64, round int) {}
```

Reason: We can't remove things like .Votes and .ProposalBlockParts from ConsensusState and call it ConsensusState -- they're vital parts of "state" just as necessary as those other fields.  And those structures aren't cheap to copy, esp for chains with larger validator sets, and the best way to mutate those things are via a pointer, w/ the intuitive interfaces for VoteSet, PartSet.  Ultimately there's a map, and maps shouldn't be copied around so lightly.

However, you can still set up the test the way you want, by making the test harness (inefficiently) copy the ConsensusState deeply before each iota operation, so you can still get the effect that you want, which I think is great.

So basically, I would prefer:

```go
func (cs *ConsensusState) RunEvent(event) ([]Message) {}
```

"Message" should be a slice.  Returning messages from `ConsensusState.***()` instead of calling eventbus methods in the ConsensusExecutor does make sense.